pkg/controller/csi_handler.go

```diff
@@ -253,14 +271,20 @@ func (h *csiHandler) csiAttach(va *storage.VolumeAttachment) (*storage.VolumeAtt
 		return va, nil, err
 	}
 
-	nodeID, err := h.getNodeID(h.attacherName, va.Spec.NodeName)
+	nodeID, err := h.getNodeID(h.attacherName, va.Spec.NodeName, nil)
 	if err != nil {
 		return va, nil, err
 	}
 
-	va, err = h.addVAFinalizer(va)
-	if err != nil {
-		return va, nil, fmt.Errorf("could not add VolumeAttachment finalizer: %s", err)
+	originalVA := va
+	va, finalizerAdded := h.prepareVAFinalizer(va)
+	va, nodeIDAdded := h.prepareVANodeID(va, nodeID)
```

Comment:
it is not clear the copy semantics here. In fact I looked through and it seems like *both* `prepare` functions do a deepcopy of the `va`. Since these two functions are both only called together here, wondering if we can wrap both in one `prepareVA` function that does the deepcopy so we are not repeating work.

Reply:
I can do `DeepCopy()` before calling `prepareVAFinalizer`, but then I do `DeepCopy` even when no change is needed. Now I do two DeepCopies when one is needed and zero copies when none is needed.

I'd like to avoid funky bool parameters to `prepareVAFinalizer` or `prepareVANodeID` if they should do the clone by themselves or it has been already cloned by previous method.

Reply:
hmm I see what you mean. I am ok with either option, feel free to revert if you think the other way is better.

Reply:
I restored double `DeepCopy()`